### PR TITLE
Disabling application elements fixed

### PR DESCRIPTION
**Pull request by pboguslawski**
Gitea does not remove all elements (UI, cron tasks) of some
disabled features which makes UI unnecessarily complicated.

This mod removes these unnecessary elements from UI/cron:

* token edition in user settings is not available
  when ENABLE_SWAGGER=false,

* application tab is not available in user settings
  if swagger and oauth2 are both disabled,

* SSH keys editor is not displayed if SSH is disabled
  and tab name changed to more general "Keys",

* resync_all_sshkeys cron task is disabled if SSH
  is disabled,

* deploy keys tab is not displayed if SSH is disabled,

* git hook access option hidden in user settings when
  git hooks are disabled,

* git hook sync cron task removed when git hooks
  are disabled.

Author-Change-Id: IB#1105071

### CODE REVIEW

**Comment by CirnoT** [comment]
Please rebase your fork, all your recent PRs are opened straight off with conflicts.

**Comment by silverwind** [comment]
There is a "API" Link in the bottom right of the footer which could also be hidden if `DisableSwagger` is true.

**Comment by lafriks** [comment]
`DISABLE_GIT_HOOKS` does not disable Gitea git hooks but task you are disabling are actually exactly for that

**Comment by silverwind** [comment]
Should consolidate existing `EnableSwagger` context variable created [here](https://github.com/go-gitea/gitea/blob/07c9f6dca4e77abedf421b14fc928323a80b1bf0/modules/context/context.go#L344) to either use the new helper or set the same `EnableSwagger` context so it's consistent.

**Comment by pboguslawski** [comment]
> DISABLE_GIT_HOOKS does not disable Gitea git hooks as they are essential for Gitea to work but task you are disabling are actually exactly for that

Is there any way to change or break git hooks in gitea repos when DISABLE_GIT_HOOKS=true (other than intentional messing from shell outside of gitea - this may be repaired with tmp setting change and running task)?

Any reason for repairing repo git hooks after initial repo creation with correct gitea hooks if DISABLE_GIT_HOOKS was always true (=should forbid any changes in git hooks by gitea)?

If newer gitea need to update its hooks after upgrade to newer gitea version it should probably do migration like in DB (migration cli?) and not require additional cron tasks.

**Comment by pboguslawski** [comment]
> but task you are disabling are actually exactly for that

Remote touching git hooks is risky so DISABLE_GIT_HOOKS should disable touching it permanently, even with cron tasks if not absolutely necessarry. If repairing / mirgation is required - should probably be done using cli during service downtime.

**Comment by pboguslawski** [comment]
Recreating hooks during gitea web downtime using gitea cli seems also safer solution to avoid races during short period when files are absent during recreation in https://github.com/go-gitea/gitea/blob/master/modules/repository/hooks.go#L56

**Comment by CirnoT** [comment]
> Any reason for repairing repo git hooks after initial repo creation with correct gitea hooks if DISABLE_GIT_HOOKS was always true (=should forbid any changes in git hooks by gitea)?

Changing configuration, moving servers and probably few other edge cases where they might get outdated.

> resync_all_sshkeys cron task is disabled if SSH
is disabled,

Might also want to hide it when `SSH_CREATE_AUTHORIZED_KEYS_FILE` is false

**Comment by pboguslawski** [comment]
> Please rebase your fork, your recent PRs are opened straight off with conflicts.

According to The Golden Rule of Rebasing...

https://www.atlassian.com/git/tutorials/merging-vs-rebasing#the-golden-rule-of-rebasing

...we do not rebase public branches. Just merged current master to this PR and resolved conflicts.

> Should consolidate existing EnableSwagger context variable created here to either use the new DisableSwagger helper or set the same EnableSwagger context so it's consistent.

Removed redundant DisableSwagger; existing EnableSwagger is used instead.

> Changing configuration, moving servers and probably few other edge cases where they might get outdated.

All sounds like admin cli tasks under downtime not cron business (see comment about race above also).

> Might also want to hide it when SSH_CREATE_AUTHORIZED_KEYS_FILE is false

Dependency added (also && !setting.SSH.StartBuiltinServer like in ssh_key.go > rewriteAllPublicKeys()). New task registerRewriteAllPrincipalKeys will be disabled also in similar way.

**Comment by pboguslawski** [comment]
Just added one more enhancement - removing external account management when openid signin/up is disabled. Not 100% sure is done right so please verify before merging.

**Comment by pboguslawski** [comment]
One mode change: 4b867e9 introduces DISABLE_2FA parameter in [security] section
of app.ini (by default set to false). If set to true it disables access to 2FA feature in
user preferences (not required in some environments i.e. when reverse proxy auth
is used). Authentication code using 2FA and any existing 2FA configuration are
left untouched.

This mod hides also security tab in user preferences when openid is also
disabled; for this reason this mod is not separate PR but exiting PR
enhancement.

File: templates/user/settings/keys_ssh.tmpl [hunk 1]
```diff
@@ -1,11 +1,8 @@
+{{if not DisableSSH}}
```
Comment (silverwind) [inline templates/user/settings/keys_ssh.tmpl:1]: Care to indent after each `if`?

**Comment by silverwind** [comment]
Would suggest not piling too many changes into one PR as it get's harder to review every time :)

**Comment by lafriks** [comment]
Why would there be need to disable 2FA to me it seems unnecessary

**Comment by pboguslawski** [comment]
> Why would there be need to disable 2FA to me it seems unnecessary

2FA in gitea does not work if header auth is enabled so its unnecessarry in this scenario; more, 2FA should be an option not requirement (i.e. reverse proxy may take care of whole auth stuff); we want to use as clean vcs UI as possible without unnecessarry stuff thats why this parameter.

BTW: please consider accepting new optional gitea features only if provided with parameter to disable it if not used.

**Comment by pboguslawski** [comment]
> Would suggest not piling too many changes into one PR as it get's harder to review every time

We don't plan to extend this PR; 2FA mod was included here because this mod hides also security tab which depends on 2FA switch.

**Comment by wxiaoguang** [comment]
Thanks for the PR. After reading the code and the comments, I feel that there are still many problems here.

1. Too many changes are mixed together, and main branch has changed a lot, many conflicts here.
2. `DISABLE_2FA` should never be used. If you do not need it, just leave it there, it won't cause any problem.
3. The non-English locale files should not be changed in PR, they are managed by crowdin automatically.
4. `EnableSwagger` is used to control the Swagger API Document, but this PR uses it for `DeleteApplication` / `loadApplicationsData`, I think it is incorrect.

I have a feeling that if some UI elements need to be hidden, we need a new and clear PR for that, instead of mixing things together. Since this PR has been `stale` for a long time, maybe we could close this one and use a new cherry-picked PR for a clear purpose.

**Comment by pboguslawski** [comment]
> Too many changes are mixed together, and main branch has changed a lot, many conflicts here

Will spit changes to separate PR with update to current main.

> `DISABLE_2FA` should never be used. If you do not need it, just leave it there, it won't cause any problem.

`DISABLE_2FA` should be used in systems where reverse proxy handles auth with strong enough solutions (2FA seems not even work when reverse proxy is enabled in gitea...) and admin don't want to leave unnecessary elements in UI. This mod was spit to #18481.

**Comment by pboguslawski** [comment]
Disabling unnecessary SSH elements was split to #18482.

**Comment by pboguslawski** [comment]
Disabling unnecessary GitHooks elements was split to  #18485.

**Comment by pboguslawski** [comment]
> EnableSwagger is used to control the Swagger API Document, but this PR uses it for DeleteApplication / loadApplicationsData, I think it is incorrect.

Thanks for this explanation; swagger part has been dropped.

**Comment by pboguslawski** [comment]
Disabling unnecessary OpenID elements was split into #18491.

**Pull request closed without merging**